### DISCUSSION TITLE
change file descriptor without re-initializing the handle of uv_poll_t type

### DISCUSSION THREAD

**Discussion by metalalive** [libuv/libuv#3587] [Q&A] [answered]
Hi , I have an application project running on Linux environment, which includes `libuv` and another third-party library, the third-party library provides APIs for starting a TCP connection to remote server (say `xxx_connect()`) and getting file descriptor of the active connection (say `xxx_get_socket()`) . So far I managed to get valid file descriptor from `xxx_get_socket()` after `xxx_connect()` completed successfully, and initialize `uv_poll_t` handle with that  file descriptor in my program.

Currently I am working on reconnecting function, after reconnecting the same server (by running `xxx_connect()` again), `xxx_get_socket()` returns different file descriptor, that means it is necessary to update `io_watcher.fd` member of a `uv_poll_t` handle to receive data in the new active connection.

AFAIK  `uv_poll_init()` internally invokes `uv__io_check_fd()` , `uv__nonblock()` and `uv__io_init()` , it seems possible to modify  `io_watcher.fd` of a  `uv_poll_t` handle **without** closing the handle and then initializing it again (see sample code below), which has extra latency. However I'm not sure if it is safe to do so,  I don't know whether `io_watcher.fd` member of a `uv_poll_t` handle is referenced elsewhere in `libuv` (e.g. `uv_run()`) which makes thing more complex. Is my approach feasible or should I re-initialize the `uv_poll_t` handle in such case ?  Appreciate any feedback.

Possible approach , simplified sample code :
```c
int uv_poll_change_fd( uv_poll_t *handle, int new_fd ) {
   if (uv__fd_exists(handle->loop, new_fd)) 
   // ..... some code ....
   err = uv__io_check_fd(handle->loop, new_fd);
   if(err)
   // ..... some code ....
   err = uv__nonblock(new_fd, 1);
   // ..... some code ....
   handle->io_watcher.fd = new_fd;
}
```

**Comment by vtjnash** [comment, marked as answer]
Would not recommend it (even though if uv_poll_stop has been called, it is guaranteed not to be accessed elsewhere until uv_poll_start). But just make a new one.

**Reply by metalalive** [reply]
Thanks for the reply
